libs/utils/energy_model.py

```diff
@@ -897,6 +995,43 @@ def read_idle_states(cpu, domain_level):
                    root_power_domain=root_pd,
                    freq_domains=freq_domains)
 
+    @classmethod
+    def from_target(cls, target):
+        """
+        Create an EnergyModel by reading a target filesystem
+
+        If present, load an EM provided via dt using from_sd_target, since these
+        devices make the full EM available via the sched domain in sysfs. If
+        there is no EM at this location, attempt to load the simplified EM
+        made available via dedicated sysfs files.
+
+        :param target: Devlib target object to read filesystem from. Must have
+                       cpufreq and cpuidle modules enabled.
+        :returns: Constructed EnergyModel object based on the parameters
+                  reported by the target.
+        """
+        _log = logging.getLogger('EMReader')
+
+        def em_present_in_sd(target):
+            def sge_path(cpu, domain, group, field):
+                f = '/proc/sys/kernel/sched_domain/cpu{}/domain{}/group{}/energy/{}'
+                return f.format(cpu, domain, group, field)
+            # Read all the files we might need in one go, otherwise this will take
+            # ages.
+            sge_globs = ['/proc/sys/kernel/sched_domain/cpu**/domain**/group**/energy/cap_states',]
+            sge_file_values = read_multiple_oneline_files(target, sge_globs)
+
+            if not sge_file_values:
+                return False
+            return True
+
+        if em_present_in_sd(target):
+            _log.info('Attempting to load EM from sched domains.')
+            return cls.from_sd_target(target)
+        else:
```

Comment:
`elif target.file_exists('/sys/devices/system/cpu/energy_model/fd0'):` ?

Reply:
Same as above - I think we only know that the fd index will be the cpuid of the first cpu in a group. If the energy_model directory is present, maybe that's enough to indicate the feature should be there and we can fail if the content is not present or not structured as expected.

Reply:
OK, I'm trying a directory check approach in the new version. Done.

Reply:
I think CPU0 is guaranteed to be the first CPU of at least one freq domain, so fd0 should always exist. But yeah, checking the directory should be just fine anyway ;-)